### PR TITLE
Buffer list filtering

### PR DESCRIPTION
**Pull request by aadcg**
This a first draft to discuss `buffer-list`.

In my view, the way I have implemented `buffer-list` is adequate and flexible.  It can merely return `(alex:hash-table-values (buffers *browser*))`; or functions, taking it as argument, are passed in such a way that the transformations are composed/chained.

This flexibility is overkill if one just wants to pass a single transformation.

Say we'd want to sort by time:

Up to know we had a nice syntax, `(buffer-list :sort-by-time t)`, but `buffer-list` was defined poorly.

With the new definition, these would be equivalent:

1)

```lisp
(sort (buffer-list) #'local-time:timestamp> :key #'last-access)
```

2)

```lisp
(buffer-list :fn-list (list (lambda (x) (sort x
                                              #'local-time:timestamp>
                                              :key #'last-access))))
```

3)

```lisp
(defun sort-by-time (x)
  (sort x #'local-time:timestamp> :key #'last-access))

(buffer-list :fn-list (list #'sort-by-time))
```

1) is by far the sanest solution, IMO.

---

Now another question comes when one wants to pass 2 or more transformations. We want something akin to 2) or 3).  In other words, does one want to define beforehand the functions passed in `:fn-list`? I'd say no.  We can do that when, and where we need it. I'd stick with 2). What sense would it make to define such functions when thinking about `buffer-list`, when they're generally defined for sequences?

The most complicated case we have is `(buffer-list :sort-by-time t :domain domain)`, which now would become:

(notice that `domain` is obviously unbound.)

```lisp
(buffer-list :fn-list (list (lambda (x) (sort x
                                              #'local-time:timestamp>
                                              :key #'last-access))
                            (lambda (x) (remove-if-not
                                         (lambda (buffer) (equal domain (quri:uri-domain (url buffer))))
                                         x))))
```

---

When we agree on something, I can propagate the changes.

@jmercouris @Ambrevar @aartaka

### CODE REVIEW

File: source/buffer.lisp [hunk 1]
```diff
@@ -667,18 +667,25 @@ If DEAD-BUFFER is a dead buffer, recreate its web view and give it a new ID."
...
-                                        buffer-list #'local-time:timestamp> :key #'last-access)
-                          buffer-list))
-         (buffer-list (if domain (remove-if-not
-                                  (lambda (i) (equal domain (quri:uri-domain (url i)))) buffer-list)
-                          buffer-list)))
-    buffer-list))
-
-(defun buffers-get (id)
-  (gethash id (slot-value *browser* 'buffers)))
+(defun buffer-list (&key (fn-list nil fn-list-supplied-p))
```
Comment (Ambrevar) [inline source/buffer.lisp:670]: Parameter names should tell the user what they do, not what type they are.  Types can be instructed in a declaim form instead, and the user can consult it from the function documentation.

So instead of `fn-list` I'd name it `filters` for instance.

Also for convenience I'd allow single functions, not necessarily list of functions.  Just wrap the argument use with `(uiop:ensure-list ...)`.

**Comment by Ambrevar** [comment]
I'm OK with this change, but I don't understand what you are asking.  Can you rephrase?

**Comment by aartaka** [comment]
I may be out of context here, but why make a list of functions instead of one function? The caller can use `alex:compose` themselves, while most frequent use-case is a single lambda one. In this situation, a single function argument (`predicate`/`filter`?) fits more than list-of-functions one.

An added benefit of implying one function is that you can use `#'identity` by default and get rid of the argument checking.

What I can definitely agree with, though, is that `buffer-list` needs more generality in sorting/filtering. Thanks for initiating the discussion, André!

**Comment by aadcg** [comment]
You do make a very good point @aartaka.

**Comment by aadcg** [comment]
And which one would you use @aartaka?

```
(buffer-list :filter (lambda (x) (sort x #'local-time:timestamp> :key #'last-access)))
(sort (buffer-list) #'local-time:timestamp> :key #'last-access)
```

**Comment by aartaka** [comment]
Most probably the second one, but that depends on the context :)

**Comment by aadcg** [comment]
Exactly. That's why I don't fully agree with you. Passing the filter/predicate key only makes sense when you need something more convoluted, which will always imply using `alex:compose`.

@Ambrevar suggestion regarding `uiop:ensure-list` is also insightful. I will draft smth.

**Comment by aadcg** [comment]
This is how I see it. 

> I don't understand what you are asking. Can you rephrase?

Look at the changes. It results in more lines then what we had before, right? Basically we need to write the same predicate several times. Do you get my questions and the difference between alternatives 2) and 3) in my initial post?

**Comment by Ambrevar** [comment]
The repeated can be extract to top-level functions, no problem.
Does this bother you?

**Comment by aadcg** [comment]
> Does this bother you?

It does.  For instance, where should a function like this one be located? 

```lisp
(defun sort-by-time (x)
  (sort x #'local-time:timestamp> :key #'last-access))
```

The other functions are even more problematic. It seems silly to write functions to be used EXCLUSIVELY for `buffer-list`. For instance, turning the following into a function has several problems. If `domain` becomes an argument, then I can't pass it to `buffer-list` with the `filter` key. Why? Because the functions that get composed are functions that only take `buffer-list`, i.e. `(alex:hash-table-values (buffers *browser*))` as argument. Any other parameters, such as `domain`, must be in the lexical scope but can't be passed directly as an argument.

```lisp
(lambda (buffer-list) (remove-if-not
                       (lambda (buffer) (equal domain (quri:uri-domain (url buffer))))
                       buffer-list))
```

**Comment by aartaka** [comment]
> The other functions are even more problematic. It seems silly to write functions to be used EXCLUSIVELY for `buffer-list`. For instance, turning the following into a function has several problems. If `domain` becomes an argument, then I can't pass it to `buffer-list` with the `filter` key. Why? Because the functions that get composed are functions that only take `buffer-list`, i.e. `(alex:hash-table-values (buffers *browser*))` as argument. Any other parameters, such as `domain`, must be in the lexical scope but can't be passed directly as an argument.
> 
> ```lisp
> (lambda (buffer-list) (remove-if-not
>                        (lambda (buffer) (equal domain (quri:uri-domain (url buffer))))
>                        buffer-list))
> ```

[`alex:curry`&`alex:rcurry`](https://common-lisp.net/project/alexandria/draft/alexandria.html#Data-and-Control-Flow)?

I'm still on the side of a more conservative&conventional filter argument, as there are tons of functions/marcos supporting this idea in e.g., Alexandria :)

**Comment by Ambrevar** [comment]
You can also use the convenient `sera:equals` to generate the predicate.

**Comment by aadcg** [comment]
You're giving me ideas on how to express the filtering, but that's not the main point. Filtering is a bad word btw. We have a list, and we apply transformation to it. We compose functions that act on it, on a certain order (and the order matters). Filtering does little justice to such a thing, but I can't think of a better word too.

The topic at hand here (for me) is: we need to rewrite calls such as `buffer-list :sort-by-time t :domain domain` without being too verbose and replicating code. Notice that as things stand in master, there's no code replication since the logic behind these keys (`sort-by-time` or `domain`) are part of `buffer-list`.


My solution thus far is:

```lisp
(buffer-list
 :filter (list
          (lambda (buffer-list) (remove-if-not
                            (lambda (buffer)
                              (equal domain
                                     (quri:uri-domain (url buffer))))
                            buffer-list))
          (lambda (buffer-list) (sort buffer-list
                                 #'local-time:timestamp>
                                 :key #'last-access))))
```

But that means I have to write all of this each time I need `buffer-list :sort-by-time t :domain domain`, which sucks. And, as mentioned above, I have doubts that writing these `lambda`s as functions is a good idea. 


Show me your solutions.

**Comment by Ambrevar** [comment]
As I mentioned, we can define these lambdas as top-level functions.

Your concern was "where to store them".  Anywhere it fits, buffer.lisp
is fine, we can even store them in a dedicated file full of these
transformers.

About the naming: true, "filter" is misleading, what about "transform" then?

**Comment by aadcg** [comment]
> As I mentioned, we can define these lambdas as top-level functions.

I have mentioned that defining a function that expects as argument `(alex:hash-table-values (buffers *browser*))` seems too much of a specific case to deserve such a treatment.

I took your suggestion and used "transform". Tho, honestly, I wouldn't even use a `key`. I would put it as an optional argument.

**Comment by Ambrevar** [comment]
> I have mentioned that defining a function that expects as argument `(alex:hash-table-values (buffers *browser*))` seems too much of a specific case to deserve such a treatment.

Can you give examples?

**Comment by aadcg** [comment]
> Can you give examples?

```lisp
(remove-if-not (match-domain domain)
               (mapcar #'url (buffer-list)))
```

This could be called with smth along the lines of `(buffer-list :transform #'match-domain-buffer-list)`.  But that seems silly, since you'd be defining a function that expects `(alex:hash-table-values (buffers *browser*))` as input. That's too much of a specific case to define as a top-level function.

**Comment by Ambrevar** [comment]
I'm confused by your last paragraph.

What could be called?  I don't see the relation between the `remove-if-not` snippet and `(buffer-list :transform #'match-domain-buffer-list)`.  Can you provide the complete code you have in mind?

Which function would expect `(alex:hash-table-values (buffers *browser*))`?

**Comment by aadcg** [comment]
> Which function would expect (alex:hash-table-values (buffers *browser*))?

The functions that you pass in the `:transform` key. They take `(alex:hash-table-values (buffers *browser*))` as argument right?

I think that if you'll look at the complete pull request you'll understand me.  Notice that I've defined `sort-by-time`, but didn't define something along those lines to match the domain. I think you'll understand why.

**Comment by Ambrevar** [comment]
Every time you said `(alex:hash-table-values (buffers *browser*))`, you meant they take a list of buffers, right?
If so, what's the problem?
The code seems completely fine to me.

The definition of `sort-by-time` seems good to me, and I don't understand why you didn't define a domain matcher.  Can you explain, I'm still confused.

Can you highlight what bothers you explicitly?

**Comment by aadcg** [comment]
> Every time you said (alex:hash-table-values (buffers *browser*)), you
> meant they take a list of buffers, right?

Right.

> If so, what's the problem?
> The code seems completely fine to me.
>
> The definition of sort-by-time seems good to me, and I don't understand
> why you didn't define a domain matcher. Can you explain, I'm still
> confused.

Because `sort-by-time` can be used generically with any sequence.  It is
true that when the key is `last-access`, then the sequence must be of
objects that have a `last-access` slot.  But anyway, `sort-by-time` is
useful in contexts other than `buffer-list` (as you can see in this
merge request).

But that wouldn't be the case for a domain matcher!  For the required
sequences are very specific ones, i.e. sequences of objects that must
have a URL slot.  So writing a function along the lines of:

```
(defun domain-matcher (x domain)
  (remove-if-not (match-domain domain)
                 (mapcar #'url x)))
```

Is utterly misleading since `x` is nothing but `(alex:hash-table-values
(buffers *browser*))`.

It could be written in the following way, tho I also find it misleading
since we would be defining a top-level function for a VERY specific
case.  So, to my mind, the only proper way to do this (minus the way
I've done it with lambdas) is to be able to define functions in the
"lexical scope" of buffer-list.  That is, is it possible to define
functions that could only be used (that is passed as argument) to
`buffer-list`?  I think that what I need is smth resembling a closure
but not quite, since I don't want to call a certain function within
`buffer-list`.  I want to say that a function can only be called when
passed as argument to `buffer-list`!

```
(defun domain-matcher (buffer-list domain)
  (remove-if-not (match-domain domain)
                 (mapcar #'url buffer-list)))
```

I hope this makes sense.

**Comment by Ambrevar** [comment]
The answer is that the matcher must return the lambda.
See the `match-domain` function that's already in the source.

So you'd write something along those lines:

```lisp
(defun match-domain (domain &rest other-domains)
  "Return a predicate for URLs matching one of DOMAIN or OTHER-DOMAINS."
  #'(lambda (buffer-or-url)
      (some (alex:curry #'string= (quri:uri-domain (url buffer-or-url))
            (cons domain other-domains))))
```

Then you can just call it this way:

```lisp
(buffer-list :transform (match-domain "foo.xyz"))
```

Does that help?

**Comment by aadcg** [comment]
Pierre Neidhardt ***@***.***> writes:

> The answer is that the matcher must return the lambda.
> See the match-domain function that's already in the source.

I'm making use of match-domain.

> So you'd write something along those lines:
>
> (defun match-domain (domain &rest other-domains)
>   "Return a predicate for URLs matching one of DOMAIN or OTHER-DOMAINS."
>   #'(lambda (buffer-or-url)
>       (some (alex:curry #'string= (quri:uri-domain (url buffer-or-url))
>             (cons domain other-domains))))
>
> Then you can just call it this way:
>
> (buffer-list :transform (match-domain "foo.xyz"))

This is not what I want.  The functions passed to buffer-list take
buffer-list, i.e. (alex:hash-table-values (buffers *browser*)), as
argument.

But I get the idea, so I could do:

```lisp
(defun domain-match (domain)
  #'(lambda (buffer-list) (remove-if-not (match-domain domain)
                                         (mapcar #'url buffer-list))))
```

and then call (buffer-list :transform (domain-match "foo.xyz")).

Seems ok to me.

**Comment by Ambrevar** [comment]
It seems we are on the same page.

In my example, I conflated match-domain to work both on URLs and buffers.
Indeed, the `url` method works on both types.

**Comment by aadcg** [comment]
Take a look @Ambrevar. Thanks.

File: source/buffer.lisp [hunk 2]
```diff
@@ -667,15 +667,25 @@ If DEAD-BUFFER is a dead buffer, recreate its web view and give it a new ID."
...
+(defun buffer-list (&optional (transform #'identity))
+  "Return the list of buffers of `*browser*'.
+TRANSFORM is a list of functions, or a single function, that take a list of
+buffers as argument.  These functions are composed from right-to-left."
+  (funcall (apply #'alex:compose (alex:ensure-list transform))
+           (alex:hash-table-values (buffers *browser*))))
+
+(defun domain-match (domain)
+  #'(lambda (buffer-list) (remove-if-not (match-domain domain)
+                                    (mapcar #'url buffer-list))))
```
Comment (Ambrevar) [inline source/buffer.lisp:679]: Having both `domain-match` and `match-domain` is confusing.
I suggest just having this in place of the existing `match-domain`:

```lisp
(defun match-domain (domain)
   #'(lambda (buffers-or-urls) (remove-if-not (match-domain domain)
                                     (mapcar #'url buffers-or-urls))))
```

File: source/buffer.lisp [hunk 3]
```diff
@@ -740,8 +750,7 @@ proceeding."
 
 (defun last-active-buffer ()
   "Return buffer with most recent `last-access'."
-  (first (sort (buffer-list)
-               #'local-time:timestamp> :key #'last-access)))
+  (first (buffer-list #'sort-by-time)))
```
Comment (Ambrevar) [inline source/buffer.lisp:753]: As was suggested, this is equivalent to `(sort-by-time (buffer-list))`, so I suggest we keep the latter which is more unversal, and we drop the `transform` argument since you deprecated it.

File: source/buffer.lisp [hunk 4]
```diff
@@ -763,7 +772,7 @@ proceeding."
   (echo "~a copied to clipboard." (title (current-buffer))))
 
 (defun buffer-initial-suggestions (&key current-is-last-p domain)
-  (let ((buffers (buffer-list :sort-by-time t :domain domain)))
+  (let ((buffers (buffer-list (list (domain-match domain) #'sort-by-time))))
```
Comment (Ambrevar) [inline source/buffer.lisp:775]: Then this would be `(funcall (match-domain domain) (sort-by-time (buffer-list))`.

Comment (Ambrevar) [reply source/buffer.lisp:775]: And similar for the other calls.

File: source/buffer.lisp [hunk 5]
```diff
@@ -862,7 +871,7 @@ set of useful URLs or preparing a list to send to a someone else."
 
 (define-command delete-all-buffers (&key confirmation-p)
   "Delete all buffers, with confirmation."
-  (let ((count (length (buffer-list))))
+  (let ((count (buffer-list #'length)))
```
Comment (Ambrevar) [inline source/buffer.lisp:874]: I prefer the former `(length (buffer-list))`, it's more idiomatic CL.

File: source/status.lisp [hunk 6]
```diff
@@ -42,15 +42,16 @@
...
-                    (mapcar #'(lambda (i) (quri:uri-domain (url i)))
-                            (buffer-list :sort-by-time t)))
-            :test #'equal)))
+  (let ((buffer-domains
+          (buffer-list (list (lambda (buffer-list)
+                               (remove-duplicates
+                                (sera:filter-map #'quri:uri-domain
+                                                 (mapcar #'url buffer-list))
+                                :test #'equal))
+                             #'sort-by-time))))
```
Comment (Ambrevar) [inline source/status.lisp:51]: We can call `match-domain` here, no?

**Comment by Ambrevar** [comment]
OK, I like where this is going.  If this is alright with everyone, I suggest dropping the `transform` argument and just use regular function calls, it's just as short and regular CL anyways.

File: source/buffer.lisp (hunk 2 above)
Comment (aadcg) [reply source/buffer.lisp:679]: > Having both domain-match and match-domain is confusing.

This is a fair concern.

> I suggest just having this in place of the existing match-domain

These are two different things anyway. I agree `match-domain` and friend could handle both a URL or a buffer, but one thing is the lambda that returns the predicate that does the matching, and other thing is passing a list and getting the list back with the matches of the filtering.

File: source/buffer.lisp (hunk 3 above)
Comment (aadcg) [reply source/buffer.lisp:753]: Fair enough.

**Comment by Ambrevar** [comment]
> These are two different things anyway. I agree `match-domain` and friends could handle both a URL or a buffer, but one thing is the lambda that returns the predicate that does the matching, and other thing is passing a list and getting the list back with the matches of the filtering.

Sorry, you are right, replace `funcall` by `remove-if` or `sera:filter` in my
suggestions, thus

```lisp
(sera:filter (match-domain "foo") (buffer-list))
```

is good enough, no?

File: source/buffer.lisp (hunk 4 above)
Comment (aadcg) [reply source/buffer.lisp:775]: But this completely defeats the way I defined `buffer-list`. This would only make sense if `buffer-list` would be nothing else but `(alex:hash-table-values (buffers *browser*))`.

How can `(funcall (match-domain domain) (sort-by-time (buffer-list))` be more sensible than `(buffer-list (list (domain-match domain) #'sort-by-time))`?  

The second syntax is plain simple. It says: take `buffer-list` as argument and pass it to the functions from the list. These functions are composed from right-to-left.

File: source/buffer.lisp (hunk 5 above)
Comment (aadcg) [reply source/buffer.lisp:874]: Also fair.

File: source/status.lisp (hunk 6 above)
Comment (aadcg) [reply source/status.lisp:51]: No. Here you want a list with all the unique domains from `buffer-list`.

**Comment by aadcg** [comment]
> I suggest dropping the transform argument and just use regular function calls, it's just as short and regular CL anyways.

I disagree. All we're discussing here are the notations below to denote composition of functions `f`, `g` and `h`. 

The left hand side notation is far superior than the one on the right. In the Lisp world, I'm just putting these functions in a list, choosing my argument (`x`) and happily composing them from right-to-left.

**Comment by Ambrevar** [comment]
André Alexandre Gomes ***@***.***> writes:

> How can `(funcall (match-domain domain) (sort-by-time (buffer-list))` be more sensible than `(buffer-list (list (domain-match domain) #'sort-by-time))`?  
>
> The second syntax is plain simple. It says: take `buffer-list` as argument and pass it to the functions from the list. These functions are composed from right-to-left.

There are two ways to do this:

- either have the transforms happen inside buffer-list;

- or have no transforms in buffer-list at all, and just do them on the
call site.

The latter has the benefit that the reader who does not know the details
of buffer-list would know immediately what going on (and does not need
to read the docstring of buffer-list to know in which order the
transforms are processed).
This approach is also less code.

This is why I'm in favour of having `buffer-list` that just returns 
`(alex:hash-table-values (buffers *browser*))`.

Can you see more
benefits to having transforms inside buffer-list?

**Comment by aadcg** [comment]
> This approach is also less code.

I have doubts about this one.  But I understood the other reasons.

> 
This is why I'm in favour of having `buffer-list` that just returns
`(alex:hash-table-values (buffers *browser*))`.

Ok, I will refactor it using this approach.

**Comment by aadcg** [comment]
Is master broken? I get 

```
The value
  #<USER-INTERNAL-BUFFER {100A944643}>
is not of type
  QURI.URI:URI
```

when I start it.

**Comment by Ambrevar** [comment]
Can you open a new issue with the full stacktrace?
(When you run it from a REPL.)

**Comment by aadcg** [comment]
> Can you open a new issue with the full stacktrace?
(When you run it from a REPL.)

I can't reproduce it anymore. Might have been some mistake on my side.

**Comment by aadcg** [comment]
I think things now stand with what you had in mind @Ambrevar. 

Thoughts? @jmercouris @aartaka

**Comment by aartaka** [comment]
Looks all good to me!

**Comment by jmercouris** [comment]
I have nothing to add. Looks OK to me!

**Comment by Ambrevar** [comment]
Thanks!
Looks good to me.  I've just tested it, and I've spotted two issues.
I've fixed them on the https://github.com/atlas-engineer/nyxt/tree/buffer-list-filtering-fixes branch.

What do you think?

**Comment by jmercouris** [comment]
Why protect against NIL? should that be the responsibility of these functions?

**Comment by Ambrevar** [comment]
Since match-domain and friends return a predicate, and predicates are
likely to be called against anything.

**Comment by jmercouris** [comment]
OK. Seems consistent with the rest of CL.

**Comment by Ambrevar** [comment]
@aadcg OK to merge the fixed branch?

**Comment by aadcg** [comment]
@ambrevar, OK to merge, thanks. What's the advantage of `sera:filter` over `apply`?

**Comment by Ambrevar** [comment]
`apply` does not do what you want here.

- sera:filter == remove-if-not.
- apply: calls the function over the arguments found in the list in last
  position.

In this case, `(apply (match-domain domain) URLS)` would fail because
the predicate returned by match-domain expects only one argument.

**Comment by Ambrevar** [comment]
`apply` can return a list, it actually returns the return value of the function in first position.

**Comment by Ambrevar** [comment]
`remove-if-not` is deprecated by CLHS.  `sera:filter` is not :)

**Comment by Ambrevar** [comment]
Merged with 645e5ab0a7a41030fd832f497810cd145b911580.  Thank you André for keeping through this :)

**Comment by aadcg** [comment]
Pierre Neidhardt ***@***.***> writes:

> Merged with 645e5ab. Thank you André for keeping through this :)

Thank you.

I finally understood why I did this silly thing of using apply.  It was
because match-domain used to me smth else and then I didn't take action!
